### ISSUE TITLE
LogstashTcpSocketAppender cannot be used in combination with AccessLogs

### ISSUE THREAD

**Issue by el-toro** [logstash/logstash-logback-encoder#59]
Access logs cannot be transferred with LogstashTcpSocketAppender because it only supports ILoggingEvents.
Logback's socket appender for access logs unfortunately sends serialized java objects.
It would be nice to have a corresponding appender for AccessEvents.

**Comment by hntd187** [comment]
Would also love to have this. Currently can't get access logs to work over a socket connection. @philsttr is there some current way to accomplish this?

**Comment by mihaiplesa** [comment]
Take a look at https://github.com/mihaiplesa/logstash-logback-encoder/commit/bb87959537430e4ce00fff4da0715aa593e0c60d

**Comment by hntd187** [comment]
@mihaiplesa I can't seem to get yours to work either.

I have everything running on localhost

```
<configuration>
        <statusListener class="ch.qos.logback.core.status.OnConsoleStatusListener" />  
        <appender name="STASH" class="net.logstash.logback.appender.LogstashAccessTcpSocketAppender">
                <remoteHost>localhost</remoteHost>
                <port>3000</port>
                <encoder class="net.logstash.logback.encoder.LogstashEncoder" />
        </appender>
        <appender-ref ref="STASH" />
</configuration>
```

My logstash config...
```
input {
  tcp {
    host => "0.0.0.0"
    type => "tomcat"
    port => 3000
    codec => "json"
  }
}
output {
  elasticsearch {
    protocol => "http"
    host => "localhost"
    port => "9200"
  }
}
```

It connects, but as soon as some access happens it closes the connection...
```
14:53:31,536 |-INFO in ch.qos.logback.access.tomcat.LogbackValve[localhost] - filename property not set. Assuming [/home/scarman/logging/apache-tomcat-7.0.57/conf/logback-access.xml]
14:53:31,591 |-INFO in ch.qos.logback.core.joran.action.StatusListenerAction - Added status listener of type [ch.qos.logback.core.status.OnConsoleStatusListener]
14:53:31,536 |-INFO in ch.qos.logback.access.tomcat.LogbackValve[localhost] - filename property not set. Assuming [/home/scarman/logging/apache-tomcat-7.0.57/conf/logback-access.xml]
14:53:31,591 |-INFO in ch.qos.logback.core.joran.action.StatusListenerAction - Added status listener of type [ch.qos.logback.core.status.OnConsoleStatusListener]
14:53:31,592 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [net.logstash.logback.appender.LogstashAccessTcpSocketAppender]
14:53:31,592 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [net.logstash.logback.appender.LogstashAccessTcpSocketAppender]
14:53:31,598 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STASH]
14:53:31,598 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STASH]
14:53:31,704 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STASH] to ch.qos.logback.access.tomcat.LogbackValve[localhost]
14:53:31,704 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STASH] to ch.qos.logback.access.tomcat.LogbackValve[localhost]
14:53:31,704 |-INFO in ch.qos.logback.access.joran.action.ConfigurationAction - End of configuration.
14:53:31,704 |-INFO in ch.qos.logback.access.joran.action.ConfigurationAction - End of configuration.
14:53:31,705 |-INFO in ch.qos.logback.access.joran.JoranConfigurator@7f5220a3 - Registering current configuration as safe fallback point
14:53:31,705 |-INFO in ch.qos.logback.access.joran.JoranConfigurator@7f5220a3 - Registering current configuration as safe fallback point
14:53:31,536 |-INFO in ch.qos.logback.access.tomcat.LogbackValve[localhost] - filename property not set. Assuming [/home/scarman/logging/apache-tomcat-7.0.57/conf/logback-access.xml]
14:53:31,591 |-INFO in ch.qos.logback.core.joran.action.StatusListenerAction - Added status listener of type [ch.qos.logback.core.status.OnConsoleStatusListener]
14:53:31,592 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [net.logstash.logback.appender.LogstashAccessTcpSocketAppender]
14:53:31,598 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STASH]
14:53:31,704 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STASH] to ch.qos.logback.access.tomcat.LogbackValve[localhost]
14:53:31,704 |-INFO in ch.qos.logback.access.joran.action.ConfigurationAction - End of configuration.
14:53:31,705 |-INFO in ch.qos.logback.access.joran.JoranConfigurator@7f5220a3 - Registering current configuration as safe fallback point

14:53:31,709 |-INFO in net.logstash.logback.appender.LogstashAccessTcpSocketAppender[STASH] - remote peer localhost:3000: connection established
14:53:31,709 |-INFO in net.logstash.logback.appender.LogstashAccessTcpSocketAppender[STASH] - remote peer localhost:3000: connection established
```

On the logstash side... 
```
Accepted connection {:client=>"127.0.0.1:60865", :server=>"0.0.0.0:3000", :level=>:debug, :file=>"logstash/inputs/tcp.rb", :line=>"166"}
config LogStash::Codecs::JSONLines/@charset = "UTF-8" {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Codecs::Line/@charset = "UTF-8" {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
Connection closed {:client=>"127.0.0.1:60865", :level=>:debug, :file=>"logstash/inputs/tcp.rb", :line=>"121"}
```

Any idea what is causing this? Everything should be able to talk to one another?

**Comment by dimas** [comment]
I created another pull request to add support for IAccessEvent logging - https://github.com/logstash/logstash-logback-encoder/pull/63

Compared to Mihai's patch, it reduces copy&paste by moving the commoon code to a base class. So hopefully it will be accepted.

(Mihai's PR does a bit more that that as it changes some other code we also need changed but as these changes are not related, we will refactor them and submit as another PR separately).

**Comment by philsttr** [comment]
Merged PR #63.  This should allow sending access logs directly to logstash over TCP.

For any other enhancements / bugs, please open a new issue / pull request.

Thanks for the contribution!
